text/0000-reseeding-rand.md

```diff
@@ -0,0 +1,682 @@
+        test::black_box(&buf);
+    })
+}
+
+#[bench]
+fn chacha(b: &mut test::Bencher) {
+    let mut rng = ChaChaRng::new_unseeded();
+    let mut buf = [0; 1000];
+    b.iter(|| {
+        rng.fill_bytes(&mut buf);
+        test::black_box(&buf);
+    })
+}
+```
+
+Results with `-O`:
+
+```
+running 2 tests
+test chacha ... bench:      4495 ns/iter (+/- 140)
+test os     ... bench:     73136 ns/iter (+/- 6551)
+
+test result: ok. 0 passed; 0 failed; 0 ignored; 2 measured
+```
+
+NB. the above (*only* 16&times; slower) is actually the case when
+`OsRng` is going to perform best: reading a large chunk of bytes in
+one go. Changing the `.fill_bytes` call to `.gen::<u32>()` makes `OsRng`
+250&times; slower than `ChaChaRng`.
+
+Relevant links for this RFC:
+
+- [C++11's `<random>` header][cpp_random]
+- [D's `std.random`][d_random]
+- ["Investigate replacing the ISAAC Rng with directly using the operating system,
+  or an eSTREAM Rng (etc.)"](https://github.com/rust-lang/rust/issues/10047)
+- ["Should RNGs really be clone?"](https://github.com/rust-lang/rust/issues/20973)
+
+[cpp_random]: http://en.cppreference.com/w/cpp/numeric/random
+[d_random]: http://dlang.org/phobos/std_random.html
+
+## Cryptographic security
+
+The only interface that will be recommended as cryptographically
+secure will be the RNG that interfaces with the operating system. This
+will be crystal clear in the documentation.
+
+# Detailed design
+
+The stable-at-1.0 parts of the module will essentially just consist of
+2 traits (plus one extension trait) and 3 RNG algorithms (plus one
+wrapper).
+
+## Traits
+
+`std::rand` will include the following generic set-up;
+
+```rust
+/// Things that can be created randomly, where the creation can be
+/// mediated by information contained in `Data`.
+trait Random<Data = FullRange> {
+    /// Create a random value of type `Self`
+    fn random<R: ?Sized + Rng>(data: Data, rng: &mut R) -> Self;
+}
+
+// Example:
+
+// create a raw u8 with no restriction.
+impl Random<FullRange> for u8 {
+    ...
+}
+
+// create a u8 between the bounds of range (same exclusivity properties as usual)
+impl Random<Range<u8>> for u8 {
+    ...
+}
+
+
+/// Random number generator.
+trait Rng: io::Reader {
+    /// Generate/fetch the next `u32` (no range restriction)
+    #[unstable]
+    fn next_u32(&mut self) -> u32 { ... }
+
+    /// Generate/fetch the next `u64` (no range restriction)
+    #[unstable]
+    fn next_u64(&mut self) -> u64 { ... }
+    /// Generate/fetch the next `f32` in the range [0,1)
+    #[unstable]
+    fn next_f32(&mut self) -> f32 { ... }
+    /// Generate/fetch the next `f64` in the range [0,1)
+    #[unstable]
+    fn next_f64(&mut self) -> f64 { ... }
+}
+
+impl<'a, T: ?Sized + Rng> Rng for &'a mut T {}
+impl<T: ?Sized + Rng> Rng for Box<T> {}
+
+/// Basic extension functionality to ensure that `Rng` is object-safe.
+trait RngExt {
+    fn gen<T, D>(&mut self, data: D) -> T where T: Random<D> { ... }
```

Comment:
Note to self: this taking `D` by value is unfortunate (unnecessarily consuming), but taking `D` by reference is also unfortunate: `rng.gen(&(0..n))`.

Possible solution, take by value but have

```rust
impl<'a, T: Random<D>, D> Random<&'a D> for T {
    fn random(data: &&'a U, rng: &mut R) -> T {
        Random::random(*data, rng)
    }
}
```

which means passing a reference will automatically work.